### PR TITLE
Kinetic Software - Some Subtle Changes - David Collins

### PR DESCRIPTION
**Pull request by davidcollinsdev**
With some obvious bits addressed, I do take on board your comment in the underline repo re: `We also know that in the future this API will connect to another database which may not be MongoDB.` - I can see some coupling of Mongo within `Kx.Availability.Data` project & ideally we would want to abstract those few bits out completely so we can drive a desired implementation upstream in the `Kx.Availability` -> ServiceCollection - to put this in practise is a bit of re-jigging I think and requires a bit of thought given the current coupling to Mongo we currently have.

One observation: this codebase is accessible publically; both the original repo & now my former one.

In summary the changes include things like:

1. Semantic renaming (Enum's / Methods)
2. Redundant spaces dropped
3. Redundant (and repeated) use of `StartTime` in the BDD test feature file
4. Variable -> Private Field
5. Dropped a non used `public` member from an interface and made the member (in it's class) `private` given its access
6. Demonstrated an example of unit tests; supporting a change I have suggested to some existing validation on the request
7. "dodgy" `true` check in the spec file resulting into a false positive
8. is `baconRoll` a valid role for the API?? Looks like a possible mistaken entry.

I havent used `MongoDB` before - some of those index's defined looked "dodgy" but it could be my lack of understanding around Mongo DB and requires a bit more research into it. At a guess, the ordering of the attributes selected for the index may want their more “primary attributes” sorted first.

Anyway, I would be interested to get your thoughts on some of my subtle changes.

All tests should **still pass**

image

p.s. I am curious how your machines perform; mine was struggling only for me to notice this `vmmem.exe` process sucking my memory; driven from docker desktop, hard process to kill too. This coupled with a more resource hungry VS 2022 📈 

image

### CODE REVIEW

File: Kx.Availability.Data.Tests/TenantTests.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    //    Mongo DB's (or any other type of mocked infrastructure)
+    public class TenantTests
+    {
+        private Mock<IHttpContextAccessor> _httpContextAccessor;
+        private Mock<HttpContext> _httpContextMock;
+        private Mock<HttpRequest> _httpRequestMock;
+
+        public TenantTests()
+        {
+            SetupMocks();
```
Comment (davidcollinsdev) [inline Kx.Availability.Data.Tests/TenantTests.cs:18]: This is generally how I lay out all my test classes; as they grow organically & say the need for things like "dummy data" are required, this constructor would grow to be something like..... (adopting a clean code fashion)

```
 public TenantTests()
 {
            SetupTestData();
            SetupMocks();
 }
```

File: Kx.Availability.Data.Tests/TenantTests.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        private void SetupMocks()
+        {
+            _httpContextAccessor = new Mock<IHttpContextAccessor>();
+            _httpContextMock = new Mock<HttpContext>();
+            _httpRequestMock = new Mock<HttpRequest>();
+
+            SetupMockBehaviour();
```
Comment (davidcollinsdev) [inline Kx.Availability.Data.Tests/TenantTests.cs:27]: Notice the split of Mock instantiation & mock behaviour; dont bundle these things into one large bloated method, clean it 🧹

File: Kx.Availability.Data.Tests/TenantTests.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            _httpContextAccessor.Setup(hca => hca.HttpContext).Returns(_httpContextMock.Object);
+        }
+
+        //DC: I have just written a basic test to cover the slightly extended validation of the Request.Path (Found in the 
+        //    Tenant class) - it was previously quite "loose" before in that it only checked for a null value; not empty.
+        //    This unit test targets both specifics and drills down to the outcome; it drove the change.
+        [Theory]
+        [InlineData(null)]
+        [InlineData("")]
+        public void TenantConstructed_WhenRequestPathisNullEmptyOrWhitespace_ThrowsBadHttpRequestException(string requestPath)
```
Comment (davidcollinsdev) [inline Kx.Availability.Data.Tests/TenantTests.cs:42]: The `OrWhitespace` part can be dropped from the test name.

I originally went `Red` -> `Green` -> `Red` to discover the `OrWhitspace` scenario was never a valid test given the HTTP Request path can never be set to a white spaced string.

I dropped the scenario `Refactor` but clearly didn’t refactor the test name!

**Comment by sara-me** [comment]
Hi @davidcollinsdev,
can you please set your fork as private or delete it?

**Comment by davidcollinsdev** [comment]
Hey Sara.

Yes of course.

Deleted - DONE 👍

Regards
David

**Comment by davidcollinsdev** [comment]
> Hi @davidcollinsdev, can you please set your fork as private or delete it?

Hey Sara

That should now be deleted.

Regards
David

**Comment by sara-me** [comment]
[like]  Sara Erdelyi reacted to your message:
________________________________

**Pull request closed without merging**